### DISCUSSION TITLE
Solved: Using TFT_eSPI with an ESP8266

### DISCUSSION THREAD

**Discussion by karstengit** [Bodmer/TFT_eSPI#2147] [Q&A - General] [answered]
I already used this wonderful library successful with an STM32 Blue-Pill.
Now i am trying to connect [this display](http://www.lcdwiki.com/2.8inch_SPI_Module_ILI9341_SKU:MSP2807#Interface_Definition) to an ESP8266-12F with an own small adapter PCB.

I am using again the benchmark test that can be found in the examples.

```
/*
 Adapted from the Adafruit and Xark's PDQ graphicstest sketch.
	Connections of the display
	1 VCC		3V3
	2 GND		GND			
	3 CS		GPIO15		
	4 Reset		RST	
	5 D/C		GPIO16
	6 MOSI (SDI)	GPIO13	
	7 SCK		GPIO14
	8 LED		GPIO0
	9 MISO (SDO)	GPIO12 (only touch and not TFT)
*/

#include "SPI.h"
#include "TFT_eSPI.h"

// Use hardware SPI
TFT_eSPI tft = TFT_eSPI();

void setup() {
	// Setting the mode of the pins
	pinMode(0, OUTPUT);			// TFT LED Light
	digitalWrite(0, 1);			// LED on
	
	tft.init();
}

void loop(void)
{
	Serial.println(F("Benchmark                Time (microseconds)"));

	uint32_t usecHaD = testHaD();
	Serial.print(F("HaD pushColor            "));
	Serial.println(usecHaD);
	delay(100);
...
```

with this User_Setup.h in the same folder

```
// Only define one driver, the other ones must be commented out
#define ILI9341_DRIVER

// For NodeMCU - use pin numbers in the form PIN_Dx where Dx is the NodeMCU pin designation
#define TFT_CS   PIN_D8  // Chip select control pin D8
#define TFT_DC   PIN_D0  // Data Command control pin
// #define TFT_RST  PIN_D4  // Reset pin (could connect to NodeMCU RST, see next line)
#define TFT_RST  -1    // Set TFT_RST to -1 if the display RESET is connected to NodeMCU RST or 3.3V

// #define TFT_BL PIN_D0  // LED back-light (only for ST7789 with backlight control pin)

#define TOUCH_CS PIN_D2     // Chip select pin (T_CS) of touch screen


#define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
#define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
#define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:-.
#define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
//#define LOAD_FONT8N // Font 8. Alternative to Font 8 above, slightly narrower, so 3 digits fit a 160 pixel TFT
#define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts

// Comment out the #define below to stop the SPIFFS filing system and smooth font code being loaded
// this will save ~20kbytes of FLASH
#define SMOOTH_FONT

#define SPI_FREQUENCY  27000000

// Optional reduced SPI frequency for reading TFT
#define SPI_READ_FREQUENCY  20000000

// The XPT2046 requires a lower SPI clock rate of 2.5MHz so we define that here:
#define SPI_TOUCH_FREQUENCY  2500000
```

It compiles, can be flashed and the UART is working, but there cannot be seen any signal on the connected lines with an oscilloscope on SCK and MOSI, the CS TFT is always Low.

What i am doing wrong?

**Comment by karstengit** [comment]
Sorry for opening as bug report - I did not see that there is an active discussion forum.
https://github.com/Bodmer/TFT_eSPI/discussions/2146

**Comment by Bodmer** [comment]
I suspect this is a pin mapping problem. The library was originally developed soley for the ESP8266 NodeMCU board and thus the default pin definitions are for that particular boaard.

Unfortunately different board suppliers map different "Dxx" pins to different GPIO, so in your instance it would be best to use direct GPIO numbers.

Just for information only,  this is the mapping from PIN_Dxx definitions to GPIO for the default:

```
// These are the pins for ESP8266 boards
//      Name   GPIO    NodeMCU      Function
#define PIN_D0  16  // GPIO16       WAKE
#define PIN_D1   5  // GPIO5        User purpose
#define PIN_D2   4  // GPIO4        User purpose
#define PIN_D3   0  // GPIO0        Low on boot means enter FLASH mode
#define PIN_D4   2  // GPIO2        TXD1 (must be high on boot to go to UART0 FLASH mode)
#define PIN_D5  14  // GPIO14       HSCLK
#define PIN_D6  12  // GPIO12       HMISO
#define PIN_D7  13  // GPIO13       HMOSI  RXD2
#define PIN_D8  15  // GPIO15       HCS    TXD0 (must be low on boot to enter UART0 FLASH mode)
#define PIN_D9   3  //              RXD0
#define PIN_D10  1  //              TXD0

#define PIN_MOSI 8  // SD1          FLASH and overlap mode
#define PIN_MISO 7  // SD0
#define PIN_SCLK 6  // CLK
#define PIN_HWCS 0  // D3

#define PIN_D11  9  // SD2
#define PIN_D12 10  // SD4
```

I think these settings may be suitable for your board if it is wired like this. Note that the SPI ports are mapped to specific GPIO pins on the ESP8266 and thus MISO, MOSI and SCLK must use those pins:

```
#define TFT_MISO 12
#define TFT_MOSI 13
#define TFT_SCLK  14
#define TFT_CS      15  // Chip select control pin
#define TFT_DC       0  // Data Command control pin
#define TFT_RST      2  // Reset pin (could connect to RST pin)
```

Note also that some pins must be at a certain level at boot time for the processor to startup and run the sketch (see comments above). Typically ready built boards bias these pins into the correct state.

Run the Test and diagnostic example "Read_User_Setup" and look at the serial monitor output to check the compiler is picking up the settings correctly. Post the output here with a copy of your setup file if you are not sure the output is correct.

**Comment by Bodmer** [comment]
I have converted this to a discussion topic.

**Comment by karstengit** [comment]
The pins assignment is double checked and correct to your assignment

```
#define PIN_D5  14  // GPIO14       HSCLK
#define PIN_D6  12  // GPIO12       HMISO

```
There is only this one additional SPI-Port and the one that can be shared with the Flash-Chip.
The Hardware-SPI seems not to be initialized, because there is no output on it.

**Reply by Bodmer** [reply]
Post the output from Read_User_Setup to be sure the compiler is picking up the right pins.

Which board have you selected in the Arduino IDE?

I would connect the TFT RST to a GPIO and define it in the setup (for test purposes) as the display may not be getting reset properly.

**Reply by Bodmer** [reply]
You should be using GPIO 13 for MOSI, that sends data and commands to the display. MISO does not need to be wired unless you are reading from the displau in your sketch.

**Comment by karstengit** [comment]
GPIO13 is used for MOSI.

There can only be found the correct User_Setup.h in /tmp/arduino_build_381780/sketch and no "Read_User_Setup" ?

**Comment by karstengit** [comment]
I assume there is another fundamental problem.
Maybe the wrong board for using this driver?

![Screenshot_20221107_165659](https://user-images.githubusercontent.com/5941550/200355934-4b43b8e5-a472-4db0-87da-de1e02b4d489.png)

Everything works fine with this settings for other sketches using SPI.

**Comment by Bodmer** [comment]
The Read_User_Setup sketch is an example here:
https://github.com/Bodmer/TFT_eSPI/tree/master/examples/Test%20and%20diagnostics/Read_User_Setup

**Comment by karstengit** [comment]
Sorry for the delay.

First I did have to find out that the ESP8266 somehow did not work correctly - so I changed it to another new one.

Now
```
[code]
TFT_eSPI ver = 2.3.4
Processor    = ESP8266
Frequency    = 80MHz
Voltage      = 3.80V
Transactions = No
Interface    = SPI
SPI overlap  = No

Display driver = 9341
Display width  = 240
Display height = 320

TFT_CS   = PIN_D8
TFT_DC   = PIN_D3
TOUCH_CS = PIN_D2

Font GLCD   loaded
Font 2      loaded
Font 4      loaded
Font 6      loaded
Font 7      loaded
Font 8      loaded
Smooth font enabled

Display SPI frequency = 27.00
Touch SPI frequency   = 2.50
[/code]

```
but no working display.

I can see signals on MOSI, SCK and CS TST now, but no signal on DC.
I changed the pin from GPIO16 / D0 to GPIO0 / D3, but it is still not working.

**Reply by Bodmer** [reply]
The voltage reported is unusually high. I would expect this to be closed to 3.3V. I suggest you check the supply voltage with a meter. Thedisplay also should be powered from 3.3V unless it has an onboard 5V to 3.3V regulator.  Check you do not have a parasitic path from 5V logic to your ESP8266 as that may be the cause of your faulty ESP8266 and can affect the display interface reliability.

The reported Read_User_Setup pins do not all correspond with your posted setup file.  For example your posted setup states:

#define TFT_DC   PIN_D0  // Data Command control pin

But read user setup reports:

TFT_DC   = PIN_D3

Check User_Setup_Select.h points to ONLY ONE setup file. Check that setup file pins are correct.

Read_User_Setup should also report the RST pin as I recommended above that you connect it for test purposes.

Clearly there are setup and wiring inconsistencies that need to be fixed.

**Comment by karstengit** [comment]
Is a local User_Setup.h in the folder of the *.ino ignored?

**Reply by Bodmer** [reply]
Yes, a setup in the sketch folder will be ignored.

**Comment by TangerineDreamer** [comment, marked as answer]
I suggest to wire the led-blacklight to 3,3v,here my setup working:

//-----------------------------------------------------------

#define ILI9341_DRIVER

//MOSI>D7,SCK>D5, MISO> better (on the air)




#define TFT_CS   PIN_D0  // SS/RS - Chip select control pin D8
#define TFT_DC   PIN_D8  // Data Command control pin
#define TFT_RST  -1    // Set TFT_RST to -1 if the display RESET is connected to NodeMCU RST or 3.3V

//#define TOUCH_CS PIN_D3     // Chip select pin (T_CS) of touch screen




#define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
#define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
#define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:-.
#define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
//#define LOAD_FONT8N // Font 8. Alternative to Font 8 above, slightly narrower, so 3 digits fit a 160 pixel TFT
#define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts


#define SMOOTH_FONT



#define SPI_FREQUENCY  20000000

#define SPI_READ_FREQUENCY  20000000


#define SPI_TOUCH_FREQUENCY  2500000
//-----------------------------------------------------------------------

**Comment by karstengit** [comment]
LED-Backlight and Reset was already fixed to 3V3.

Now I tried exact your configuration (copyCATandpasteDOG) and it really works!
Thank you very much for your support!

But why it does not work with
```
#define TFT_CS   PIN_D8  // Chip select control pin D8
#define TFT_DC   PIN_D3  // Data Command control pin
#define TFT_RST  -1    // Set TFT_RST to -1 if the display RESET is connected to NodeMCU RST or 3.3V
```
?
Pins are not interchangeable?

**Reply by Bodmer** [reply]
Pins are interchangeable but some pins have dual functions as posted above and duplicated here:
```
// These are the pins for ESP8266 boards
//      Name   GPIO    NodeMCU      Function
#define PIN_D0  16  // GPIO16       WAKE
#define PIN_D1   5  // GPIO5        User purpose
#define PIN_D2   4  // GPIO4        User purpose
#define PIN_D3   0  // GPIO0        Low on boot means enter FLASH mode
#define PIN_D4   2  // GPIO2        TXD1 (must be high on boot to go to UART0 FLASH mode)
#define PIN_D5  14  // GPIO14       HSCLK
#define PIN_D6  12  // GPIO12       HMISO
#define PIN_D7  13  // GPIO13       HMOSI  RXD2
#define PIN_D8  15  // GPIO15       HCS    TXD0 (must be low on boot to enter UART0 FLASH mode)
#define PIN_D9   3  //              RXD0
#define PIN_D10  1  //              TXD0

#define PIN_MOSI 8  // SD1          FLASH and overlap mode
#define PIN_MISO 7  // SD0
#define PIN_SCLK 6  // CLK
#define PIN_HWCS 0  // D3

#define PIN_D11  9  // SD2
#define PIN_D12 10  // SD4
```
PIN_D3 (GPIO 0) has a dual function so if your display board pulls it low at boot time, viz after sketch upload (maybe via a pull down or parasitic bias) then the processor will not boot and enters FLASH mode so the sketch will not run. I suspect this is what is happening in this case. NodeMCU boards have a pullup on that pin so use of that GPIO is safe.

**Reply by karstengit** [reply]
Such a thing with PIN_D3 (GPIO 0) seems to have happen.
My board has no pullup on this pin.
I have tested the first ESP8266 again now and it is working.

**Comment by TangerineDreamer** [comment]
It's a mistery,i had to search all over the net to discover this by hazard this setup,there is no reason for me why it's not working either,
i think that comes from the esp8266,but esp32 accepts more possibilities.
Have you tried to interchange the pins?

**Comment by karstengit** [comment]
Thanks for confirming this "bug".
Up to now I first must check if I can use the touchpad too.

I have no idea how the pin mapping is done?

![mapping](https://user-images.githubusercontent.com/5941550/200677210-0f454d04-d43d-450c-8d87-a41d68b92301.png)

Here in the library the NodeMCU pins are used, but normally the Arduino pin numbers can be used.
This Arduino pins will be mapped internally again. That is complete nonsense.

**Reply by Bodmer** [reply]
It is confusing because the Arduino team did not get involved with the ESP8266 board package, it was a community effort, this means different board suppliers use different Dx to GPIO mappings and hence sketch portability between boards is lost. These mappings are in the variants folder of the board package [here](https://github.com/esp8266/Arduino/tree/master/variants). The NodeMCU format dominated the market for a while so board makers like Wemos changed to the NodeMCU mappings which confused customers even more for a while!

The selection of "Generic ESP8266 Module" means that the "Arduino" pin references D0 etc are not even defined! So for example:
```
 pinMode(D0, OUTPUT);
```
Will generate a compile error. If other boards are selected then D0 is usually defined, but may be allocated to different GPIO so it does get confusing! So for the "Generic ESP8266 Module" where the PCB does not have pins labelled Dxx it is best to use the GPIO number as that is probably what the number means on the board. Some boards use IOxx which implies this is a GPIO reference.

This means that it is usually best to use the correct board module in the IDE to get the Dx allocations to GPIO correct for the Dx board markings.

So how to wire touch:

T_CLK to same GPIO as TFT SCK
T_CS to TOUCH_CS GPIO
T_DIN to same GPIO as TFT SDI(MOSI)
T_DO to same GPIO as TFT SDO(MISO)
T_IRQ not connected

**Reply by karstengit** [reply]
> These mappings are in the variants folder of the board package [here](https://github.com/esp8266/Arduino/tree/master/variants).

The mapping seems to map only to arduino pins and only for some of them.

.../arduino15/packages/esp8266/hardware/esp8266/3.0.1/variants/generic/common.h

```
#define PIN_SPI_SS   (15)
#define PIN_SPI_MOSI (13)
#define PIN_SPI_MISO (12)
#define PIN_SPI_SCK  (14)

static const uint8_t SS    = PIN_SPI_SS;
static const uint8_t MOSI  = PIN_SPI_MOSI;
static const uint8_t MISO  = PIN_SPI_MISO;
static const uint8_t SCK   = PIN_SPI_SCK;

#ifndef PIN_A0
#define PIN_A0 (17)
#endif /* PIN_A0 */

```

The complete mapping seems to be a mystic procedure.
I searched but could not find an header file that is doing it.

**Reply by Bodmer** [reply]
You will not find a mapping of Dx to GPIO for the generic module. Generic modules that are not products for use in the Arduino environment do not have the pins labelled Dx on the PCB, so in this case it is expected that GPIO numbers are used directly to identify the digital inputs and outputs.

**Reply by karstengit** [reply]
But it is compiled now with "Generic ESP8266 module" and it works.
So the mystic pinmapping works here too.

**Comment by TangerineDreamer** [comment]
Hey karstengit,my setup i gave you works also for touch,just uncomment:
//#define TOUCH_CS PIN_D3 // Chip select pin (T_CS) of touch screen

**Reply by karstengit** [reply]
Thanks.
Here it works with
#define TOUCH_CS PIN_D2     // Chip select pin (T_CS) of touch screen
I am using the example Keypad_240x320.ino
